onnxruntime/core/providers/cuda/cuda_allocator.cc

```diff
@@ -77,6 +77,20 @@ void* CUDAExternalAllocator::Alloc(size_t size) {
 
 void CUDAExternalAllocator::Free(void* p) {
   free_(p);
+  std::lock_guard<OrtMutex> lock(lock_);
+  auto it = reserved_.find(p);
+  if (it != reserved_.end()) {
+    reserved_.erase(it);
+    if (empty_cache_) empty_cache_();
+  }
+}
+
+void* CUDAExternalAllocator::Reserve(size_t size) {
+  void* p = Alloc(size);
+  std::lock_guard<OrtMutex> lock(lock_);
+  ORT_ENFORCE(reserved_.find(p) == reserved_.end());
```

Comment:
nit: I understand that this is just a sanity check but it seems like if ORT_ENFORCE threw, the `p` leaks. 
Also, should there be a check on whether `p` is a nullptr and an appropriate error message if that was the case ?

Reply:
Added nullptr check. For the sanity check, I just don't it's not possible the set contains p in current useage. And the Reserve() in bfc_arena.cc also has this same problem.

Reply:
Sure, all right.

Reply:
At least for current usage it's thread-safe, as this CUDAExternalAllocator only use PyTorch's CUDACachingAllocator's alloc and free, which already handle the multi-thread case.